chore(samples): environment variable connection strings

YoDaMa opened the pull request:
To prevent any accidental leaking of credentials and enforce better coding practices, this PR replaces all instances in samples where we ask a user to paste in their connection string. Instead it takes in either an environment variable named IOTHUB_CONNECTION_STRING, or a pasted connection string if the user so desires. 

There are also updates to the samples document. 

1) I have moved the list of samples below the instructions (since in my use of the readme, I have always automatically skipped over it, so it's not intuitive to have it at the top). 
2) I have tweaked some of the instructions including adding some instructions for VSCode. While writing this piece I found out we have a gist fild that contains the VSCode outline, but considering I've been in the team for months and not known it was there I'm calling that not intuitively accessible. I think this change will make our documentation more smooth for users.

Finally, there's a section of the setup readme that says the samples are not linked to the locally built folder. That's not true if you build the samples in the sample folder, so that is updated.

pierreca commented:
2 suggestions:
1. process.env.IOTHUB_CONNECTION_STRING works too (and i think looks a little better than the ["..."] notation - also: linter may barf which might annoy customers who run them
2. if the good practice we're pushing is to use env variables then we shouldn't offer the "|| your connection string here" at all. it usually is more confusing to beginners when there are 2 ways to do things instead of a single recommended way. It'll also make the readme clearer.

YoDaMa commented:
/azp run node-canary

YoDaMa commented:
I've addressed the feedback in both points. 

> 2 suggestions:
> 
>     1. process.env.IOTHUB_CONNECTION_STRING works too (and i think looks a little better than the ["..."] notation - also: linter may barf which might annoy customers who run them

Replaced with the process.env.IOTHUB_CONNECTION_STRING format.


>     2. if the good practice we're pushing is to use env variables then we shouldn't offer the "|| your connection string here" at all. it usually is more confusing to beginners when there are 2 ways to do things instead of a single recommended way. It'll also make the readme clearer.

removed the second option and removed reference to it in the readme.

anthonyvercolano commented on device/samples/device_blob_upload.js:
```diff
@@ -7,10 +7,10 @@ var Protocol = require('azure-iot-device-mqtt').Mqtt;
-var deviceConnectionString = '[Device Connection String]';
-var filePath = '[path/to/file]';
+var connectionString = process.env.IOTHUB_CONNECTION_STRING;
+var filePath = "<YourPathToFile>";
```
>"<YourPathToFile>" [](start = 15, length = 18)

Why double quotes?

anthonyvercolano commented on device/samples/device_methods.js:
```diff
@@ -10,15 +10,9 @@ var Client = require('azure-iot-device').Client;
     // open a connection to the device
-    var deviceConnectionString = process.argv[2];
-    client = Client.fromConnectionString(deviceConnectionString, Protocol);
+    var connectionString = process.env.IOTHUB_CONNECTION_STRING;
```
>process.env.IOTHUB_CONNECTION_STRING [](start = 27, length = 36)

Typically this has referred to the connection string used for the hub not for a device.

anthonyvercolano commented on device/samples/dmpatterns_fwupdate_device.js:
```diff
@@ -7,13 +7,7 @@ var Protocol = require('azure-iot-device-mqtt').Mqtt;
-}
-
-var connectionString = process.argv[2];
+var connectionString = process.env.IOTHUB_CONNECTION_STRING;
```
>IOTHUB_CONNECTION_STRING [](start = 35, length = 24)

And how does going from command line argument to environment variable improve things.
Also see previous comment about the use of IOTHUB_CONNECTION_STRING.

anthonyvercolano commented on device/samples/dmpatterns_reboot_device.js:
```diff
@@ -5,13 +5,8 @@
-}
 
-var connectionString = process.argv[2];
+var connectionString = process.env.IOTHUB_CONNECTION_STRING;
```
>process.env.IOTHUB_CONNECTION_STRING [](start = 23, length = 36)

symbol inappropriate.

anthonyvercolano commented on device/samples/edge_downstream_device.js:
```diff
@@ -18,7 +18,7 @@ var Message = require('azure-iot-device').Message;
 // The resulting string should look like the following
 //  "HostName=<iothub_host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>;GatewayHostName=<edge device hostname>"
-var connectionString = '[Downstream device IoT Edge connection string]';
+var connectionString = process.env.IOTHUB_CONNECTION_STRING;
```
> process.env.IOTHUB_CONNECTION_STRING [](start = 22, length = 37)

symbol inappropriate.

anthonyvercolano commented on device/samples/readme.md:
```diff
@@ -2,6 +2,104 @@
+
+This should install the proper dependencies as specified in the **package.json** you copied.
+
+**3) Set the IOTHUB_CONNECTION_STRING environment variable**
```
>IOTHUB_CONNECTION_STRING [](start = 13, length = 24)

symbol inappropriate.

anthonyvercolano commented on device/samples/readme.md:
```diff
@@ -2,6 +2,104 @@
+   "env": {
+      "DEBUG": "mqtt*,rhea*,azure*",
+      "NODE_ENV": "development",
+      "IOTHUB_CONNECTION_STRING": "<YourConnectionString>"
```
>IOTHUB_CONNECTION_STRING [](start = 7, length = 24)

symbol inappropriate.

anthonyvercolano commented on device/samples/readme.md:
````diff
@@ -2,6 +2,104 @@
+   ]
+}
+```
+* Edit the `IOTHUB_CONNECTION_STRING` environment variable to use your IoT Hub device connection string.
````
>IOTHUB_CONNECTION_STRING [](start = 12, length = 24)

symbol inappropriate.

anthonyvercolano commented on device/samples/simple_sample_device_with_sas.js:
```diff
@@ -13,7 +13,7 @@ var Message = require('azure-iot-device').Message;
 //  "SharedAccessSignature sr=<iothub_host_name>/devices/<device_id>&sig=<signature>&se=<expiry>"
-var sas = "[SharedAccessSignature]";
+var sas = process.env["IOTHUB_SAS"] || "<YourIoTHubSharedAccessSignature>";
 
```
Fix.

anthonyvercolano commented on device/ts-samples/sample_device.ts:
```diff
@@ -6,7 +6,11 @@ import { Mqtt as Protocol } from 'azure-iot-device-mqtt';
+
+// Connection String can either be passed in as a parameter, set as an environment 
+// variable, or manually entered.
+const connectionString = process.env["IOTHUB_CONNECTION_STRING"] || process.argv[2] || "<YourIoTHubConnectionString>";
```
> || process.argv[2] || "<YourIoTHubConnectionString> [](start = 64, length = 52)

Remove

anthonyvercolano commented on service/samples/configuration_sample.js:
```diff
@@ -5,7 +5,7 @@
 var iothub = require('azure-iothub');
 
-var connectionString = '[IoT Hub Connection String]';
+var deviceConnectionString = process.env.DEVICE_CONNECTION_STRING;
```
> deviceConnectionString = process.env.DEVICE_CONNECTION_STRING [](start = 3, length = 62)

This should be using the IOTHUB_CONNECTION_STRING.

anthonyvercolano commented on service/samples/create_device_with_cert.js:
```diff
@@ -5,7 +5,7 @@
 var iothub = require('azure-iothub');
 
-var connectionString = '<IOTHUB CONNECTION STRING>';
+var deviceConnectionString = process.env.DEVICE_CONNECTION_STRING;
```
>deviceConnectionString = process.env.DEVICE_CONNECTION_STRING; [](start = 4, length = 62)

Fix.

anthonyvercolano commented on device/samples/edge_downstream_device.js:
```diff
@@ -18,13 +18,13 @@ var Message = require('azure-iot-device').Message;
+var deviceConnectionString = process.env.DEVICE_CONNECTION_STRING;
 
 // Path to the Edge "owner" root CA certificate
 var edge_ca_cert_path = '[Path to Edge CA certificate]';
```
>'[Path to Edge CA certificate]' [](start = 24, length = 31)

Maybe this should be an environment variable too?

anthonyvercolano commented on device/samples/device_blob_upload.js:
```diff
@@ -7,8 +7,8 @@ var Protocol = require('azure-iot-device-mqtt').Mqtt;
-var deviceConnectionString = '[Device Connection String]';
-var filePath = '[path/to/file]';
+var deviceConnectionString = process.env.DEVICE_CONNECTION_STRING;
+var filePath = '<YourPathToFile>';
```
>filePath = '<YourPathToFile>' [](start = 4, length = 29)

Should this be an environment variable?

anthonyvercolano commented on device/samples/dmpatterns_fwupdate_device.js:
```diff
@@ -7,14 +7,8 @@ var Protocol = require('azure-iot-device-mqtt').Mqtt;
-
-var connectionString = process.argv[2];
-var client = Client.fromConnectionString(connectionString, Protocol);
+var deviceConnectionString = process.env.DEVICE_CONNECTION_STRING;
```
>deviceConnectionString [](start = 4, length = 22)

Did we need to introduce this single use constant?

anthonyvercolano commented on device/samples/simple_sample_device_x509.js:
```diff
@@ -14,13 +14,13 @@ var Message = require('azure-iot-device').Message;
+var deviceConnectionString = process.env.DEVICE_CONNECTION_STRING;
 var certFile = '<PATH-TO-CERTIFICATE-FILE>';
 var keyFile = '<PATH-TO-KEY-FILE>';
 var passphrase = '<KEY PASSPHRASE IF ANY>';
```
One could ask the question should there be any of these other items that should be replaced with an environment variable.
passphrase?

Also, for x509 connection strings, the are NO secrets in that connection string.  Should it have been made an environment variable then?

anthonyvercolano commented on doc/node-devbox-setup.md:
````diff
@@ -88,23 +88,30 @@ Please note that running tests and end-to-end tests require having an Azure IoT
+Then, add the environment variable `DEVICE_CONNECTION_STRING` to your terminal session. Depending on the sample, you may have to add more credentials such as a storage key. You can open the sample file to check.
+
+```bash
+$ export DEVICE_CONNECTION_STRING="<YourIoTHubConnectionString>"
````
>YourIoTHubConnectionString [](start = 36, length = 26)

It's not an IoT Hub connection string anymore...

YoDaMa commented:
/azp run node-canary
